warehouse/accounts/services.py

```diff
@@ -631,6 +633,71 @@ def get_password_timestamp(self, user_id):
         user = self.get_user(user_id)
         return user.password_date.timestamp() if user.password_date is not None else 0
 
+    def needs_tos_update(
+        self, user_id, revision, ignore_flashed=False, ignore_notified=False
+    ):
+        query = self.db.query(UserTermsOfServiceEngagement).filter(
+            UserTermsOfServiceEngagement.user_id == user_id,
+            UserTermsOfServiceEngagement.revision == revision,
+        )
+        engagements = [
+            TermsOfServiceEngagement.Viewed,
+            TermsOfServiceEngagement.Agreed,
+        ]
+        if not ignore_flashed:
+            engagements.append(TermsOfServiceEngagement.Flashed)
+        if not ignore_notified:
+            engagements.append(TermsOfServiceEngagement.Notified)
+
+        first_engagement_all = (
+            query.filter(
+                UserTermsOfServiceEngagement.datetime
+                < datetime.datetime.now(datetime.UTC) - datetime.timedelta(days=30)
+            )
```

Comment:
Is the idea here that we would re-notify the user after 30 days?

Reply:
no, we stop bothering people with the banner on login 30 days after they first saw it.